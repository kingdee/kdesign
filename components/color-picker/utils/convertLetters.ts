export function toLowerCase(str: string): string {
  let result = ''
  for (let i = 0; i < str.length; i++) {
    const charCode = str.charCodeAt(i)
    if (charCode >= 65 && charCode <= 90) {
      result += String.fromCharCode(charCode + 32)
    } else {
      result += str[i]
    }
  }
  return result
}
