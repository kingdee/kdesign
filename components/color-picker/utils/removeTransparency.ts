import { ColorTypes } from '../interface'
import { colorFormat } from './colorFormat'

export type ICurrentColorType = Exclude<typeof ColorTypes[number], 'themeColor'>

export const removeTransparency = (color: string, currentColorType: ICurrentColorType) => {
  return colorFormat(color, 1, currentColorType, true) as string
}
