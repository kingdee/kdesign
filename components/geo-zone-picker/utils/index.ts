import { RegionArray } from '../interface'

export const generateRegionMap = (letterArray: string[][], regionArray: RegionArray) => {
  const regionMap: Record<string, RegionArray> = {}

  letterArray.forEach((group) => {
    const key = group.join('') // 使用 join 将数组转换为字符串，作为键
    const matchingItems = regionArray
      .filter((item) => group.includes(item.group!.toUpperCase()))
      .sort((a, b) => a.group!.localeCompare(b.group!)) // 按照 A-Z 的顺序排序

    if (matchingItems.length > 0) {
      regionMap[key] = matchingItems
    }
  })

  return regionMap
}
