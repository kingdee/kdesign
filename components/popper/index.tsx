import React, {
  useEffect,
  useRef,
  useState,
  ReactElement,
  Children,
  forwardRef,
  useContext,
  isValidElement,
  useImperativeHandle,
} from 'react'
import { createPopper, Instance, Modifier, Placement } from '@popperjs/core'
import { tuple } from '../_utils/type'
import classnames from 'classnames'
import debounce from 'lodash/debounce'
import ReactDOM from 'react-dom'
import { isFragment } from '../_utils/reactNode'
import { ConfigContext } from '../config-provider'
import { useMergedState } from '../_utils/hooks'
import useId from '../_utils/useId'

export const Placements = tuple(
  'top',
  'left',
  'right',
  'bottom',
  'topLeft',
  'topRight',
  'bottomLeft',
  'bottomRight',
  'leftTop',
  'leftBottom',
  'rightTop',
  'rightBottom',
)

export type PlacementType = typeof Placements[number]

export const Triggers = tuple('hover', 'focus', 'click', 'contextMenu')

export type TriggerType = typeof Triggers[number]

export type RenderFunction = () => React.ReactNode

export type PopperProps = {
  gap?: number
  arrow?: boolean
  visible?: boolean
  prefixCls?: string
  arrowSize?: number
  disabled?: boolean
  arrowOffset?: number
  scrollHidden?: boolean
  mouseEnterDelay?: number
  mouseLeaveDelay?: number
  defaultVisible?: boolean
  autoPlacement?: boolean
  className?: string
  style?: React.CSSProperties
  popperClassName?: string
  popperStyle?: React.CSSProperties
  placement?: PlacementType
  tip?: any
  trigger?: TriggerType | Array<TriggerType>
  strategy?: 'fixed' | 'absolute'
  clickToClose?: boolean
  onTrigger?: (trigger: TriggerType) => void
  onVisibleChange?: (visible: boolean) => void
  getTriggerElement?: (locatorNode: HTMLElement) => HTMLElement
  getPopupContainer?: (locatorNode: HTMLElement) => HTMLElement
  onTransitionEnd?: (e: React.TransitionEvent) => void
  onAnimationEnd?: (e: React.AnimationEvent) => void
  children?: React.ReactNode
}

const useEnhancedEffect = typeof window !== 'undefined' ? React.useLayoutEffect : React.useEffect

export interface TriggerContextProps {
  registerSubPopup: (id: string, node: any) => void
}

const TriggerContext = React.createContext<TriggerContextProps | null>(null)

const generateGetBoundingClientRect = (x = 0, y = 0, width = 0, height = 0) => {
  return () => ({
    width,
    height,
    top: y,
    right: x,
    bottom: y,
    left: x,
  })
}

const virtualElement = {
  getBoundingClientRect: generateGetBoundingClientRect(),
}

const triggerTypeArray = ['hover', 'focus', 'click', 'contextMenu']

export const popperPlacementMap = {
  top: 'top',
  left: 'left',
  right: 'right',
  bottom: 'bottom',
  topLeft: 'top-start',
  topRight: 'top-end',
  bottomLeft: 'bottom-start',
  bottomRight: 'bottom-end',
  leftTop: 'left-start',
  leftBottom: 'left-end',
  rightTop: 'right-start',
  rightBottom: 'right-end',
}

const placementFlip: any = {
  bottom: 'top',
  top: 'bottom',
  right: 'left',
  left: 'right',
}

const getRealPlacement = (key: PlacementType) => {
  return popperPlacementMap[key] ? (popperPlacementMap[key] as Placement) : 'top'
}

const getFallbackPlacements: (key: Placement) => Placement[] = (key: Placement) => {
  const prefix = key.split('-')[0]
  const suffix = [''] // , '-start', '-end']

  return [
    ...[...Object.keys(placementFlip).filter((r) => [prefix, placementFlip[prefix]].includes(r))]
      .map((d) => suffix.map((o) => d + o))
      .flat()
      .filter((f) => f !== key),
    // ...[...Object.keys(placementFlip).filter((r) => ![prefix, placementFlip[prefix]].includes(r))]
    //   .map((d) => suffix.map((o) => d + o))
    //   .flat(),
  ] as Placement[]
}

const getRealDom = (locatorRef: any, locatorElement: any = undefined) => {
  if (!locatorRef.current) return locatorRef.current
  if (locatorRef.current.tagName) return locatorRef.current
  if (locatorElement) {
    const REF_NAME_OBJ: any = {
      Input: 'input',
      InputNumber: 'input',
      Select: 'select',
      Upload: 'input',
    }
    const name = REF_NAME_OBJ?.[locatorElement?.type?.displayName]
    return locatorRef?.current[name]
  }
  return locatorElement
}

const getElement = (element: any) => {
  return typeof element === 'function' ? element() : element
}

export const Popper = forwardRef<unknown, PopperProps>((props, ref) => {
  const { getPrefixCls, prefixCls: pkgPrefixCls } = React.useContext(ConfigContext)
  const {
    prefixCls,
    onTrigger,
    style,
    popperStyle,
    arrow = false,
    onVisibleChange,
    className,
    popperClassName,
    tip,
    disabled = false,
    trigger = 'click',
    placement = 'top',
    strategy = 'absolute',
    visible,
    arrowSize = 4.25,
    gap: defaultGap = 4,
    scrollHidden = false,
    mouseEnterDelay = 0.1,
    mouseLeaveDelay = 0.1,
    defaultVisible = false,
    autoPlacement = true,
    clickToClose = true,
    getTriggerElement = (locatorNode) => locatorNode,
    getPopupContainer = () => document.body,
    onTransitionEnd,
    onAnimationEnd,
    children,
  } = props

  const popperPrefixCls = getPrefixCls!(pkgPrefixCls, 'popper')
  const referencePrefixCls = `${popperPrefixCls}-reference`
  const child: any = getElement(children)
  const childrenInner = isValidElement(child) && !isFragment(child) ? child : <span>{child}</span>
  const popperElement = getElement(tip)
  const referenceElement: any = Children.only(childrenInner) as ReactElement

  const id = useId()
  const parentContext = useContext(TriggerContext)
  const subPopupRefs = useRef<Record<string, any>>({})
  const context = React.useMemo<TriggerContextProps>(() => {
    return {
      registerSubPopup: (id, subPopupEle) => {
        subPopupRefs.current[id] = subPopupEle

        parentContext?.registerSubPopup(id, subPopupEle)
      },
    }
  }, [parentContext])

  const popperRefDom = useRef<any>(null)
  const popperRefInner = useRef<any>(null)
  const popperRef: any = ref || popperRefInner

  const popperInstance = useRef<Instance | null>(null)
  const referenceRefInner = useRef<any>(null)

  const referenceRef = referenceElement?.ref || referenceRefInner
  const container = getPopupContainer(getRealDom(referenceRef, referenceElement) || document.body) || document.body

  const [visibleInner, setVisibleInner] = useMergedState<boolean>(false, {
    value: visible,
    defaultValue: defaultVisible,
  })
  const [exist, setExist] = useState<boolean>(visibleInner)
  const [placementInner, setPlacementInner] = useState<Placement>(getRealPlacement(placement))

  const delayRef = useRef<any>()
  const clearDelay = () => {
    if (typeof delayRef.current !== 'undefined') {
      clearTimeout(delayRef.current)
      delayRef.current = null
    }
  }
  const changeVisible = (nextOpen: boolean, triggerType = '') => {
    if (visibleInner !== nextOpen) {
      if (nextOpen && triggerTypeArray.includes(triggerType)) {
        onTrigger?.(triggerType as TriggerType)
      }
      if (typeof visible === 'undefined') {
        setVisibleInner(nextOpen)
      }
      onVisibleChange?.(nextOpen)
    }
    if (!nextOpen && Object.keys(subPopupRefs.current || {}).length) {
      console.log('triggerOpenSub', subPopupRefs.current)
      Object.values(subPopupRefs.current).forEach((d: any) => {
        if (typeof d?.triggerOpen === 'function' && d?.visible) {
          d?.triggerOpen(false)
        }
      })
    }
  }
  const triggerOpen = (nextOpen: boolean, triggerType = '', delay = 0) => {
    clearDelay()
    if (!disabled) {
      if (delay === 0) {
        changeVisible(nextOpen, triggerType)
      } else {
        if (visibleInner !== nextOpen) {
          delayRef.current = setTimeout(() => {
            changeVisible(nextOpen, triggerType)
          }, delay * 1000)
        }
      }
    }
  }

  const isSubPopper = (event: MouseEvent) => {
    if (!event || !event?.target) {
      return false
    }

    // let targetElement: HTMLElement = event.target as HTMLElement
    // const POP_ATTR_NAME = 'data-popper-placement'
    // while (targetElement && targetElement !== document.documentElement) {
    //   if (targetElement?.getAttribute(POP_ATTR_NAME) && targetElement?.className.includes(popperPrefixCls)) {
    //     return true
    //   }
    //   targetElement = targetElement?.parentNode as HTMLElement
    // }

    const target: HTMLElement = event.target as HTMLElement
    if (subPopupRefs.current) {
      for (const key in subPopupRefs.current) {
        const { dom } = subPopupRefs.current[key]
        if (dom && (dom?.contains(target) || dom === target)) {
          return true
        }
      }
    }

    return false
  }

  const onTriggerInner = (nextOpen: boolean, triggerType: TriggerType, delay: undefined | number = undefined) => {
    triggerOpen(nextOpen, triggerType, delay)
  }

  const onClick = (e: Event) => {
    e.stopPropagation()
    onTriggerInner(!visibleInner, 'click')
    referenceElement?.props?.onClick?.(e)
  }

  const onFocus = (e: Event) => {
    onTriggerInner(true, 'focus')
    referenceElement?.props?.onFocus?.(e)
  }

  const onBlur = (e: Event) => {
    onTriggerInner(false, 'focus')
    referenceElement?.props?.onBlur?.(e)
  }

  const onContextMenu = (e: MouseEvent) => {
    e.preventDefault()
    let clientWidth = 0
    let clientHeight = 0
    if (arrow) {
      if (placementInner.startsWith('top') || placementInner.startsWith('bottom')) {
        if (!['top', 'bottom'].includes(placementInner)) {
          clientWidth = 6 * arrowSize
        }
      } else {
        if (!['right', 'left'].includes(placementInner)) {
          clientHeight = 6 * arrowSize
        }
      }
    }

    virtualElement.getBoundingClientRect = generateGetBoundingClientRect(
      e.clientX,
      e.clientY,
      clientWidth,
      clientHeight,
    )
    onTriggerInner(!visibleInner, 'contextMenu')
    referenceElement?.props?.onContextMenu?.(e)
  }

  const onMouseOver = (e: Event) => {
    onTriggerInner(true, 'hover', mouseEnterDelay)
    referenceElement?.props?.onMouseOver?.(e)
  }

  const onMouseLeave = (e: Event) => {
    onTriggerInner(false, 'hover', mouseLeaveDelay)
    referenceElement?.props?.onMouseLeave?.(e)
  }

  const onPopperAnimationEnd = (e: React.AnimationEvent | React.TransitionEvent) => {
    onAnimationEnd?.(e as React.AnimationEvent)
    onTransitionEnd?.(e as React.TransitionEvent)
  }

  const isTrigger = (triggerValue: TriggerType) => {
    return trigger === triggerValue || (Array.isArray(trigger) && trigger.includes(triggerValue))
  }

  const triggerEventHandle = (triggerNode: any, type = 'addEventListener') => {
    if (isTrigger('click')) {
      triggerNode?.[type]('click', onClick)
    }
    if (isTrigger('focus')) {
      triggerNode?.[type]('focus', onFocus)
      triggerNode?.[type]('blur', onBlur)
    }
    if (isTrigger('contextMenu')) {
      triggerNode?.[type]('contextmenu', onContextMenu)
    }
    if (isTrigger('hover')) {
      triggerNode?.[type]('mouseover', onMouseOver)
      triggerNode?.[type]('mouseleave', onMouseLeave)
    }
  }

  useEffect(() => {
    setPlacementInner(getRealPlacement(placement))
  }, [placement])

  useEffect(() => {
    const scrollHandle = debounce((e: Event) => {
      if (visibleInner) {
        const isPopper = e.target === popperRefDom.current || popperRefDom.current?.contains?.(e.target)
        if (scrollHidden && !isPopper) {
          triggerOpen(false, 'scroll')
        }
      }
    }, 10)

    if (visibleInner) {
      document.addEventListener('scroll', scrollHandle, true)
    }

    return () => {
      document.removeEventListener('scroll', scrollHandle, true)
    }
  }, [visibleInner, scrollHidden, popperRefDom])

  useEffect(() => {
    const clickHandle = debounce((e: MouseEvent) => {
      if (visibleInner) {
        const isPopper = popperRefDom.current
          ? popperRefDom.current === e.target || popperRefDom.current.contains?.(e.target)
          : false

        const domReference = getRealDom(referenceRef, referenceElement)
        const isReference = domReference ? domReference === e.target || domReference?.contains?.(e.target) : false
        const isTarget = trigger === 'contextMenu' ? isPopper : isPopper || isReference
        console.log('clickOutside', children, e.target, isPopper, isReference, subPopupRefs.current)
        if (clickToClose && !isTarget && !isSubPopper(e)) {
          triggerOpen(false, 'clickOutside', 0)
        }
      }
    }, 10)

    if (visibleInner) {
      document.addEventListener('click', clickHandle, true)
    }

    return () => {
      document.removeEventListener('click', clickHandle, true)
    }
  }, [visibleInner, clickToClose, referenceRef, popperRefDom])

  useEffect(() => {
    const realDom = getRealDom(referenceRef, referenceElement)
    const triggerNode = getTriggerElement(realDom)

    triggerEventHandle(triggerNode)

    return () => {
      triggerEventHandle(triggerNode, 'removeEventListener')
    }
  }, [getTriggerElement, visibleInner, referenceElement, referenceRef, trigger])

  useImperativeHandle(popperRef, () => {
    return {
      dom: popperRefDom.current,
      triggerOpen,
      visible: visibleInner,
    }
  })

  const popperModifiers: Partial<Modifier<any, any>>[] = [
    {
      name: 'offset',
      options: {
        offset: [0, defaultGap + (arrow ? 5 : 0)],
      },
    },
    {
      name: 'preventOverflow',
      enabled: autoPlacement,
      options: {
        mainAxis: true,
      },
    },
    {
      name: 'flip',
      enabled: autoPlacement,
      options: {
        fallbackPlacements: autoPlacement ? getFallbackPlacements(placementInner as any) : undefined,
      },
    },
    {
      name: 'onUpdate',
      enabled: true,
      phase: 'afterWrite',
      fn: (d) => {
        const { placement: p } = d?.state
        if (p !== placementInner) {
          setPlacementInner(p)
        }
      },
    },
  ]

  const popperOptionsInner: any = {
    placement: placementInner,
    modifiers: popperModifiers,
    strategy,
  }

  useEnhancedEffect(() => {
    if (visibleInner) {
      if (!exist) {
        setExist(true)
      } else {
        if (popperInstance.current) {
          popperInstance.current?.setOptions((options) => ({
            ...options,
            ...popperOptionsInner,
          }))
          popperInstance.current?.forceUpdate()
        }
        setTimeout(() => {
          parentContext?.registerSubPopup(id, popperRef.current)
        }, 10)
      }
    }
  }, [visibleInner, placementInner, arrow])

  useEnhancedEffect(() => {
    if (!exist) {
      return undefined
    }

    const current = getRealDom(referenceRef, referenceElement)

    if (current) {
      popperInstance.current = createPopper(
        trigger === 'contextMenu' ? virtualElement : current?.closest(`.${referencePrefixCls}`) || current,
        popperRefDom.current,
        popperOptionsInner,
      )
      setTimeout(() => {
        parentContext?.registerSubPopup(id, popperRef.current)
      }, 10)
    }

    return () => {
      popperInstance.current?.destroy()
    }
  }, [exist])

  if (children === null || typeof children === 'undefined') {
    return null
  }

  const arrowStyle: Record<string, string> = {
    [`--arrowSize`]: arrowSize + 'px',
  }

  const popperContainerProps = {
    ref: popperRefDom,
    style: { ...(arrow ? arrowStyle : {}) },
    className: classnames(popperPrefixCls, { hidden: !visibleInner }),
  }

  const popperProps = {
    className: classnames(
      [`${popperPrefixCls}-${placementInner}`],
      { [`${popperPrefixCls}-${placementInner}-out`]: !visibleInner },
      {
        [`${popperPrefixCls}-${placementInner}-in`]: visibleInner,
      },
      prefixCls,
      popperClassName,
      className,
    ),
    style: { ...popperStyle, ...style },
    onMouseOver: trigger === 'hover' ? () => onTriggerInner(true, 'hover', mouseEnterDelay) : undefined,
    onMouseLeave: trigger === 'hover' ? () => onTriggerInner(false, 'hover', mouseLeaveDelay) : undefined,
    onAnimationEnd: onPopperAnimationEnd,
    onTransitionEnd: onPopperAnimationEnd,
  }

  const referenceProps = {
    ref: referenceRef,
    className: classnames(referenceElement?.props?.className, referencePrefixCls),
  }

  return (
    <>
      {React.cloneElement(referenceElement, referenceProps)}
      {exist &&
        container &&
        ReactDOM.createPortal(
          <TriggerContext.Provider value={context}>
            <div {...popperContainerProps}>
              <div {...popperProps}>
                {popperElement}
                {arrow && <div className="arrow" data-popper-arrow="" />}
              </div>
            </div>
          </TriggerContext.Provider>,
          container,
        )}
    </>
  )
})

Popper.displayName = 'Popper'

export default Popper
