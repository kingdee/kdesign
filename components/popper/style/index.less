@import '../../style/themes/index';
@import '../../style/mixins/index';

@popper-prefix-cls: ~'@{kd-prefix}-popper';

.@{popper-prefix-cls} {
  opacity: 1;
  z-index: @z-index-popper;
  visibility: visible;

  &.hidden {
    opacity: 0;
    visibility: hidden;
    pointer-events: none;
  }

  .arrow {
    visibility: hidden;

    &::before {
      visibility: visible;
      position: absolute;
      border-style: solid;
      border-color: transparent;
      border-width: var(--arrowSize);
      content: '';
      transform: rotate(45deg);
    }
  }

  &[data-popper-placement^='top'] .arrow {
    bottom: calc(1.1 * var(--arrowSize));

    &::before {
      left: calc(-1 * var(--arrowSize));
      border-top-color: transparent;
      border-left-color: transparent;
    }
  }

  &[data-popper-placement^='bottom'] .arrow {
    top: calc(-0.8 * var(--arrowSize));

    &::before {
      left: calc(-1 * var(--arrowSize));
      border-bottom-color: transparent;
      border-right-color: transparent;
    }
  }

  &[data-popper-placement^='left'] .arrow {
    right: calc(1.1 * var(--arrowSize));

    &::before {
      top: calc(-1 * var(--arrowSize));
      border-left-color: transparent;
      border-bottom-color: transparent;
    }
  }

  &[data-popper-placement^='right'] .arrow {
    left: calc(-0.8 * var(--arrowSize));

    &::before {
      top: calc(-1 * var(--arrowSize));
      border-top-color: transparent;
      border-right-color: transparent;
    }
  }
}

.popper-motion(@className, @keyframeName, @duration: calc(@transition-duration - 0.1s)) {
  .@{className} {
    animation-duration: @duration;

    &-in {
      animation-name: ~'@{keyframeName}In';
      animation-timing-function: @ease-out;
    }

    &-out {
      animation-name: ~'@{keyframeName}Out';
      animation-timing-function: @ease-in;
    }
  }
}

.popper-motion(~'@{popper-prefix-cls}-top', kdZoomBottom);
.popper-motion(~'@{popper-prefix-cls}-left', kdZoomRight);
.popper-motion(~'@{popper-prefix-cls}-right', kdZoomLeft);
.popper-motion(~'@{popper-prefix-cls}-bottom', kdZoomTop);
.popper-motion(~'@{popper-prefix-cls}-top-start', kdZoomLeftBottom);
.popper-motion(~'@{popper-prefix-cls}-right-end', kdZoomLeftBottom);
.popper-motion(~'@{popper-prefix-cls}-top-end', kdZoomRightBottom);
.popper-motion(~'@{popper-prefix-cls}-left-end', kdZoomRightBottom);
.popper-motion(~'@{popper-prefix-cls}-left-start', kdZoomTopRight);
.popper-motion(~'@{popper-prefix-cls}-bottom-end', kdZoomTopRight);
.popper-motion(~'@{popper-prefix-cls}-right-start', kdZoomTopLeft);
.popper-motion(~'@{popper-prefix-cls}-bottom-start', kdZoomTopLeft);


