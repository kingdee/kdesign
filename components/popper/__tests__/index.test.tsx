import React from 'react'
import { mount } from 'enzyme'
import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom'
import Popper from '../index'
import Input from '../../input'
import { popperPlacementMap } from '../../popper'
import mountTest from '../../../tests/shared/mountTest'
import { sleep } from '../../../tests/utils'
import { act } from 'react-dom/test-utils'
window.MutationObserver = require('mutation-observer')

describe('Popper test visible', () => {
  beforeEach(() => {
    ;(window as any).document.getSelection = jest.fn(() => {
      return {
        removeAllRanges: jest.fn(),
      }
    })
  })

  it('calls onVisibleChange when visibility is toggled', async () => {
    let visible = false
    const onVisibleChange = jest.fn((v) => {
      visible = v
    })

    const user = userEvent.setup()
    const { container } = render(
      <Popper onVisibleChange={onVisibleChange} trigger={'click'} tip={<div>tip</div>}>
        <button>Content</button>
      </Popper>,
    )
    const btn = await screen.findByText('Content')

    await user.click(btn)
    await sleep(300)
    expect(await screen.findByText('tip')).toBeTruthy()
    expect(visible).toEqual(true)
    expect(container).toMatchSnapshot()
    expect(onVisibleChange).toHaveBeenCalled()
    await user.click(btn)
    await sleep(300)
    expect(visible).toBe(false)
  })
})

describe('Popper', () => {
  // 1. mount test
  mountTest(Popper)

  // 2. render test
  it('render Test', () => {
    const wrapper = mount(
      <Popper tip={<span>text</span>}>
        <span>text</span>
      </Popper>,
    )
    expect(wrapper).toMatchSnapshot()
  })
  // 3. warns in components
  // 4. render null or undefined without errors
  it('render null or undefined without errors', () => {
    expect(
      mount(
        <Popper>
          {null}
          {undefined}
        </Popper>,
      ),
    ).toMatchSnapshot()
  })

  // 5. displayName
  it('should have displayName static property', () => {
    const wrapper = mount(
      <Popper tip="tips text">
        <span>text</span>
      </Popper>,
    )
    expect((wrapper.type() as any).displayName).toBe('Popper')
  })

  // 6. API test
  describe('API test', () => {
    it('arrow is useful', () => {
      const wrapperRef: any = React.createRef()
      const wrapper = mount(
        <div ref={wrapperRef}>
          <Popper tip="test" arrow defaultVisible getPopupContainer={() => wrapperRef.current}>
            <span>text</span>
          </Popper>
        </div>,
      )
      expect(wrapper.find('.kd-popper').find('[data-popper-arrow]')).toHaveClassName('arrow')
    })

    it('defaultVisible & visible', async () => {
      const wrapper = mount(
        <Popper defaultVisible visible={false} tip="test" arrow trigger={'click'}>
          <div className={'kd-test'}>text</div>
        </Popper>,
      )

      expect(wrapper.find('.kd-popper').length).toEqual(0)
      wrapper.setProps({ visible: true })
      wrapper.update()
      expect(wrapper.find('.kd-popper').length).toEqual(1)
    })

    it('getTriggerElement is useful', async () => {
      const ref: any = React.createRef()
      const triggerRef: any = React.createRef()
      const wrapper = mount(
        <div ref={ref}>
          <button ref={triggerRef}>test</button>
          <Popper
            defaultVisible
            getTriggerElement={() => triggerRef.current}
            tip="tips text"
            getPopupContainer={() => ref.current}
          >
            <span className="kd-test-trigger">text</span>
          </Popper>
        </div>,
      )
      triggerRef.current.click()
      await sleep(300)
      expect(wrapper).toMatchSnapshot()
    })

    it('popperClassName & popperStyle is useful', () => {
      const ref: any = React.createRef()
      const wrapper = mount(
        <Popper
          tip="text"
          defaultVisible
          popperStyle={{ color: 'red' }}
          popperClassName="test"
          getPopupContainer={() => ref.current}
        >
          <span>text</span>
        </Popper>,
      )
      expect(wrapper.find('.kd-popper-top')).toHaveClassName('test')
      expect(wrapper.find('.kd-popper-top')).toHaveStyle({ color: 'red' })
    })

    it('className & placement is useful', () => {
      Object.entries(popperPlacementMap).forEach(([key, value]: any) => {
        const wrapper = mount(
          <Popper tip="text" defaultVisible placement={key} className={'my-popper'}>
            <span>text</span>
          </Popper>,
        )
        expect(wrapper.find('.my-popper').at(1)).toHaveClassName(`kd-popper-${value}`)
      })
    })

    it('trigger Popper', async () => {
      const content = 'tips text'
      const ref = React.createRef() as any
      const wrapper = mount(
        <div ref={ref}>
          <Popper tip={content} defaultVisible getPopupContainer={() => ref.current}>
            <span className="kd-test-trigger">text</span>
          </Popper>
        </div>,
      )

      wrapper.find('.kd-test-trigger').simulate('mouseleave')
      await sleep(300)
      expect(wrapper.find('.kd-popper').length).toBeTruthy()
    })
  })

  // 7. poperClass popperStyle
  it('should poperClass popperStyle useful', () => {
    const ref: any = React.createRef()
    const wrapper = mount(
      <div ref={ref}>
        <Popper
          tip="test"
          defaultVisible
          popperClassName="test"
          popperStyle={{ color: 'red' }}
          getPopupContainer={() => ref.current}
        >
          <div>test</div>
        </Popper>
      </div>,
    )
    expect(wrapper.find('.kd-popper-top')).toHaveClassName('test')
    expect(wrapper.find('.kd-popper-top')).toHaveStyle({ color: 'red' })
  })

  // 8. component interaction(event)
  it('component interaction(event)', async () => {
    const onTrigger = jest.fn()
    const onVisibleChange = jest.fn()
    const ref: any = React.createRef()
    const wrapper = mount(
      <Popper ref={ref} defaultVisible tip="tip text" onVisibleChange={onVisibleChange} onTrigger={onTrigger}>
        <div className={'kd-test-event'}>text</div>
      </Popper>,
    )
    expect(wrapper.find('.kd-test-event').length).toEqual(1)
    await act(() => {
      ref.current.triggerOpen(false)
    })
    expect(onVisibleChange).toBeCalled()
    await act(() => {
      ref.current.triggerOpen(true, 'hover', 0)
    })
    expect(onTrigger).toBeCalled()
  })

  // 9. config Provider
  // 10. ref test
  it('ref  is useful ', async () => {
    const ref: any = React.createRef()
    const wrapper = mount(
      <Popper ref={ref} defaultVisible tip="tip text" trigger={'click'}>
        <div className={'kd-test'}>text</div>
      </Popper>,
    )
    expect(wrapper.find('.kd-test').length).toEqual(1)
    ref.current.triggerOpen(false)
    await sleep(100)
    expect(ref.current.visible).toEqual(false)
    expect(wrapper).toMatchSnapshot()
  })

  // 11. special case
  describe('special case', () => {
    it('测试ref返回值不为dom元素', () => {
      const wrapper = mount(
        <Popper tip="tips text">
          <Input borderType="bordered" prefix="KDesign" suffix="Kdesign" />
        </Popper>,
      )
      expect(wrapper).toMatchSnapshot()
    })
  })
})
