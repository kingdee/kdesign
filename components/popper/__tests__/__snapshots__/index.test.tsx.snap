// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popper API test getTriggerElement is useful 1`] = `
<div>
  <button>
    test
  </button>
  <Popper
    defaultVisible={true}
    getPopupContainer={[Function]}
    getTriggerElement={[Function]}
    tip="tips text"
  >
    <span
      className="kd-test-trigger kd-popper-reference"
    >
      text
    </span>
    <Portal
      containerInfo={
        <div>
          <button>
            test
          </button>
          <span
            class="kd-test-trigger kd-popper-reference"
          >
            text
          </span>
          <div
            class="kd-popper hidden"
          >
            <div
              class="kd-popper-top kd-popper-top-out"
            >
              tips text
            </div>
          </div>
        </div>
      }
    >
      <div
        className="kd-popper"
        style={Object {}}
      >
        <div
          className="kd-popper-top kd-popper-top-in"
          onAnimationEnd={[Function]}
          onTransitionEnd={[Function]}
          style={Object {}}
        >
          tips text
        </div>
      </div>
    </Portal>
  </Popper>
</div>
`;

exports[`Popper render Test 1`] = `
<Popper
  tip={
    <span>
      text
    </span>
  }
>
  <span
    className="kd-popper-reference"
  >
    text
  </span>
</Popper>
`;

exports[`Popper render null or undefined without errors 1`] = `
<Popper>
  <span
    className="kd-popper-reference"
  />
</Popper>
`;

exports[`Popper special case 测试ref返回值不为dom元素 1`] = `
<Popper
  tip="tips text"
>
  <Input
    borderType="bordered"
    className="kd-popper-reference"
    prefix="KDesign"
    suffix="Kdesign"
  >
    <ClearableInput
      borderType="bordered"
      className="kd-popper-reference"
      element={
        <input
          className="kd-input kd-input-size-middle"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyUp={[Function]}
          prefix="KDesign"
          type="text"
          value=""
        />
      }
      focused={false}
      handleReset={[Function]}
      inputCount={null}
      inputType="input"
      prefix="KDesign"
      prefixCls="kd-input"
      size="middle"
      suffix="Kdesign"
      type="text"
      value=""
    >
      <span
        className="kd-input-wrapper kd-input-wrapper-size-middle kd-popper-reference"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <span
          className="kd-input-prefix"
          onMouseDown={[Function]}
        >
          KDesign
        </span>
        <input
          className="kd-input kd-input-size-middle"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyUp={[Function]}
          prefix="KDesign"
          style={null}
          type="text"
          value=""
        />
        <span
          className="kd-input-suffix"
          onMouseDown={[Function]}
        >
          Kdesign
        </span>
      </span>
    </ClearableInput>
  </Input>
</Popper>
`;
